### DISCUSSION TITLE
`upgrades` and `prBodyDefinitions` access in templates

### DISCUSSION THREAD

**Discussion by amezin** [renovatebot/renovate#29546] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Bitbucket Server, Docker image ghcr.io/renovatebot/renovate:37.388.1-full

### Please tell us more about your question or problem

This is likely two separate issues, but initially I'll start it as one discussion because of the context.

So, I have a config preset, where dependencies are associated with Jira issues using `packageRules` and `prBodyDefinitions`. 

Because of how Bitbucket Server <-> Jira integration works, issues need to be referenced in the pull request title and commit message to link with the PR and commit. I can't simply use `commitMessagePrefix`, because updates can be grouped - so they should have multiple issue references, all in the commit message and PR title, and `commitMessagePrefix` isn't "mergeable".

But if I try to use this: `"commitMessagePrefix": "{{#each upgrades}}{{{prBodyDefinitions.Issue}}} {{/each}}"` in group configuration, I get an empty prefix.

It seems that `upgrades` aren't available in the commit message/PR title at all, because `"commitMessageTopic": "{{#each upgrades}}{{depName}} {{/each}}"` also gives an empty topic.

And `prBodyDefinitions` give no result even in PR description body: `"prHeader": "Issues: {{#each upgrades}}{{{prBodyDefinitions.Issue}}} {{/each}}"` gives just `Issues: `. `upgrades`, however, are available in PR description: `"prHeader": "depNames: {{#each upgrades}}{{{depName}}} {{/each}}"` works as I expected.

But I need both `upgrades` and `prBodyDefinitions` in `commitMessagePrefix` template. Can these limitations be removed?

Reproduction: https://github.com/amezin/renovate-template-issues

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by amezin** [comment]
Is unavailability of `upgrades` (for commit title/message) and `prBodyDefinitions` (for any templates) expected, or not?

**Reply by rarkins** [reply]
Per https://docs.renovatebot.com/templates/ `upgrades` should be available. `prBodyDefinitions` was never intended to be an input for templates so no surprise but I guess there's no harm in allowing it too

**Reply by amezin** [reply]
> `prBodyDefinitions` was never intended to be an input for templates

https://docs.renovatebot.com/templates/#exposed-config-options

> The following configuration options are passed through for templating: .... [prBodyDefinitions](https://docs.renovatebot.com/configuration-options/#prbodydefinitions), ...

**Reply by rarkins** [reply]
Then you'll need to debug locally to work out why it's not being passed through

**Comment by rarkins** [comment]
It sounds like the better solution here would be that commitMessagePrefix is calculated "lazily" based on the branch data and not based on the "eager" calculation which is done today per-upgrade

**Reply by amezin** [reply]
Moving `commitMessage` and `prTitle` calculation out of per-upgrade loop seems to work, all tests pass. It makes sense (since there's only one commit and PR) and is, theoretically, even a performance improvement. 
https://github.com/renovatebot/renovate/pull/29955

It can break something if someone used per-upgrade `prTitle` and/or `commitMessage` in templates though (but, on the other hand, these templates were from group settings anyway).

**Reply by amezin** [reply]
I mean, if someone did `{{#each upgrades}} - {{{ prTitle }}}\n{{\each}}` in `prHeader`, for example, the output can change. But that `{{{ prTitle }}}` is `group.prTitle` - so different prTitles for different upgrades should be unlikely.
